### PR TITLE
Optimize using StringBuilder

### PR DESCRIPTION
**Pull request by SergerGood**

### CODE REVIEW

**Comment by ArminShoeibi** [comment]
Nice.
https://stackoverflow.com/questions/42721009/performance-of-stringbuilder-append

**Comment by roji** [comment]
@SergerGood any specific reason to do this change? Are you seeing performance differences between appending a char and a string?

@ArminShoeibi your link seems to point to a Java-related question, not .NET.

File: src/EFCore.Relational/Metadata/IViewColumn.cs [hunk 1]
```diff
@@ -44,12 +44,12 @@ string ToDebugString(MetadataDebugStringOptions options = MetadataDebugStringOpt
             var singleLine = (options & MetadataDebugStringOptions.SingleLine) != 0;
             if (singleLine)
             {
-                builder.Append($"ViewColumn: {Table.Name}.");
+                builder.Append("ViewColumn: ").Append(Table.Name).Append('.');
```
Comment (roji) [inline src/EFCore.Relational/Metadata/IViewColumn.cs:47]: This makes the code less readable, in a codepath that is not perf-sensitive in any way.

**Comment by ArminShoeibi** [comment]
> @ArminShoeibi your link seems to point to a Java-related question, not .NET.

oh, my bad.

**Comment by SergerGood** [comment]
> @SergerGood any specific reason to do this change? Are you seeing performance differences between appending a char and a string?

@roji https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1834

File: src/EFCore.Relational/Metadata/IViewColumn.cs (hunk 1 above)
Comment (SergerGood) [reply src/EFCore.Relational/Metadata/IViewColumn.cs:47]: In the same place:

https://github.com/dotnet/efcore/blob/ac2bb48b10ecf1289b568a94b7a35e8075c6d787/src/EFCore.Relational/Metadata/IViewColumn.cs#L50

https://github.com/dotnet/efcore/blob/ac2bb48b10ecf1289b568a94b7a35e8075c6d787/src/EFCore.Relational/Metadata/IViewColumn.cs#L52

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs [hunk 2]
```diff
@@ -289,13 +289,13 @@ public virtual string Identifier(string name, ICollection<string>? scope = null)
 
             if (partStart != name.Length)
             {
-                builder.Append(name.Substring(partStart));
+                builder.Append(name, partStart, name.Length - partStart);
```
Comment (roji) [inline src/EFCore.Design/Design/Internal/CSharpHelper.cs:292]: Is there any reason to do this change? The previous code seems simpler.

File: src/EFCore.Relational/Metadata/IColumn.cs [hunk 3]
```diff
@@ -158,12 +158,12 @@ string ToDebugString(MetadataDebugStringOptions options = MetadataDebugStringOpt
             var singleLine = (options & MetadataDebugStringOptions.SingleLine) != 0;
             if (singleLine)
             {
-                builder.Append($"Column: {Table.Name}.");
+                builder.Append("Column: ").Append(Table.Name).Append('.');
```
Comment (roji) [inline src/EFCore.Relational/Metadata/IColumn.cs:161]: Please revert

File: src/EFCore.Relational/Metadata/IFunctionColumn.cs [hunk 4]
```diff
@@ -44,12 +44,12 @@ string ToDebugString(MetadataDebugStringOptions options = MetadataDebugStringOpt
             var singleLine = (options & MetadataDebugStringOptions.SingleLine) != 0;
             if (singleLine)
             {
-                builder.Append($"FunctionColumn: {Table.Name}.");
+                builder.Append("FunctionColumn: ").Append(Table.Name).Append('.');
```
Comment (roji) [inline src/EFCore.Relational/Metadata/IFunctionColumn.cs:47]: Please revert

File: src/EFCore.Relational/Metadata/ISqlQueryColumn.cs [hunk 5]
```diff
@@ -44,12 +44,12 @@ string ToDebugString(MetadataDebugStringOptions options = MetadataDebugStringOpt
             var singleLine = (options & MetadataDebugStringOptions.SingleLine) != 0;
             if (singleLine)
             {
-                builder.Append($"SqlQueryColumn: {Table.Name}.");
+                builder.Append("SqlQueryColumn: ").Append(Table.Name).Append('.');
```
Comment (roji) [inline src/EFCore.Relational/Metadata/ISqlQueryColumn.cs:47]: Please revert

File: src/EFCore.Relational/Metadata/IViewColumn.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Metadata/IViewColumn.cs:47]: Still doesn't seem like an improvement to me (note that here there are three things, whereas in your two examples there are two). Can you please revert unless there's a good reason to do this?

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs [hunk 6]
```diff
@@ -706,10 +706,10 @@ public override DataTable GetSchemaTable()
                         command.Parameters.Clear();
                         var columnType = "typeof(\"" + columnName.Replace("\"", "\"\"") + "\")";
                         command.CommandText = new StringBuilder()
-                            .AppendLine($"SELECT {columnType}")
-                            .AppendLine($"FROM \"{tableName}\"")
-                            .AppendLine($"WHERE {columnType} != 'null'")
-                            .AppendLine($"GROUP BY {columnType}")
+                            .Append("SELECT ").AppendLine(columnType)
```
Comment (roji) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs:709]: Please revert

**Review by roji — CHANGES_REQUESTED** [review]
Here's a quick benchmark:

|       Method |     Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |---------:|----------:|----------:|------:|------:|------:|----------:|
|   AppendChar | 5.923 ns | 0.0093 ns | 0.0072 ns |     - |     - |     - |         - |
| AppendString | 6.682 ns | 0.1175 ns | 0.0981 ns |     - |     - |     - |         - |

So there's roughly a half-nanosecond difference between these two methods, which doesn't seem significant in any of the code paths being touched. Not all quality rules make sense in all contexts...

We can still do this, but please revert the other changes which remove string interpolation - IMO these just make the code less readable for no reason.

<details>
<summary>Benchmark code</summary>

```c#
[MemoryDiagnoser]
public class Program
{
    private StringBuilder _builder;

    [Benchmark]
    public void AppendChar()
    {
        _builder.Clear();
        _builder.Append('x');
    }

    [Benchmark]
    public void AppendString()
    {
        _builder.Clear();
        _builder.Append("X");
    }

    [GlobalSetup]
    public void Setup() => _builder = new StringBuilder();

    static void Main(string[] args)
        => BenchmarkRunner.Run<Program>();
}
```

File: src/EFCore.Relational/Metadata/IViewColumn.cs (hunk 1 above)
Comment (SergerGood) [reply src/EFCore.Relational/Metadata/IViewColumn.cs:47]: I don't know if it's good reason or not.

``` ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i5-9300H CPU 2.40GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=6.0.100-preview.3.21202.5
  [Host]     : .NET Core 6.0.0 (CoreCLR 6.0.21.20104, CoreFX 6.0.21.20104), X64 RyuJIT
  DefaultJob : .NET Core 6.0.0 (CoreCLR 6.0.21.20104, CoreFX 6.0.21.20104), X64 RyuJIT
```
|              Method | CharactersCount |     Mean |    Error |   StdDev |   Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------- |---------------- |---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| **StringInterpolation** |               **5** | **32.56 ns** | **0.679 ns** | **1.764 ns** | **32.28 ns** | **0.0134** |     **-** |     **-** |      **56 B** |
|              Append |               5 | 14.08 ns | 0.309 ns | 0.356 ns | 14.13 ns |      - |     - |     - |         - |
| **StringInterpolation** |              **10** | **35.03 ns** | **0.714 ns** | **1.046 ns** | **35.00 ns** | **0.0153** |     **-** |     **-** |      **64 B** |
|              Append |              10 | 14.56 ns | 0.903 ns | 2.502 ns | 13.31 ns |      - |     - |     - |         - |

<details>
  <summary>Benchmark code</summary>
  
  ```csharp
    [MemoryDiagnoser]
    public class Program
    {
        private StringBuilder _builder;
        private string _text;
    
        [Params(5, 10)]
        public int CharactersCount { get; set; }
    
        [Benchmark]
        public void StringInterpolation()
        {
            _builder.Clear();
            _builder.Append($"Column: {_text}.");
        }
    
        [Benchmark]
        public void Append()
        {
            _builder.Clear();
            _builder.Append("Column: ").Append(_text).Append('.');
        }
    
        [GlobalSetup]
        public void Setup()
        {
            _builder = new StringBuilder();
            _text = new string('a', CharactersCount);
        }
    
        private static void Main() => BenchmarkRunner.Run<Program>();
    }
  ```
</details>

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 2 above)
Comment (SergerGood) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:292]: The reason is that there is a ready-made native better method for this.

``` ini
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i5-9300H CPU 2.40GHz, 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=6.0.100-preview.3.21202.5
  [Host]     : .NET Core 6.0.0 (CoreCLR 6.0.21.20104, CoreFX 6.0.21.20104), X64 RyuJIT
  DefaultJob : .NET Core 6.0.0 (CoreCLR 6.0.21.20104, CoreFX 6.0.21.20104), X64 RyuJIT
```
|          Method | CharactersCount |     Mean |    Error |   StdDev |   Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------- |---------------- |---------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| **StringSubstring** |               **5** | **23.03 ns** | **1.147 ns** | **3.308 ns** | **21.55 ns** | **0.0076** |     **-** |     **-** |      **32 B** |
| AppendSubstring |               5 | 11.99 ns | 0.134 ns | 0.119 ns | 11.96 ns |      - |     - |     - |         - |
| **StringSubstring** |              **10** | **22.92 ns** | **0.341 ns** | **0.302 ns** | **22.81 ns** | **0.0076** |     **-** |     **-** |      **32 B** |
| AppendSubstring |              10 | 11.77 ns | 0.216 ns | 0.169 ns | 11.78 ns |      - |     - |     - |         - |


<details>
  <summary>Benchmark code</summary>
  
  ```csharp
    [MemoryDiagnoser]
    public class Program
    {
        private StringBuilder _builder;
        private string _text;
        private int _partStart;
    
        [Params(5, 10)]
        public int CharactersCount { get; set; }
    
        [Benchmark]
        public void StringSubstring()
        {
            _builder.Clear();
            _builder.Append(_text.Substring(_partStart));
        }
    
        [Benchmark]
        public void AppendSubstring()
        {
            _builder.Clear();
            _builder.Append(_text, _partStart, _text.Length - _partStart);
        }
    
        [GlobalSetup]
        public void Setup()
        {
            _builder = new StringBuilder();
            _text = new string('a', CharactersCount);
            _partStart = CharactersCount / 2;
        }
    
        private static void Main() => BenchmarkRunner.Run<Program>();
    }
  ```
</details>

File: src/EFCore.Relational/Metadata/IViewColumn.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Metadata/IViewColumn.cs:47]: String interpolation indeed has a perf cost, but this code is completely non-perf-sensitive, so readability is more important than 20ns.

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:292]: https://github.com/dotnet/efcore/pull/24755#discussion_r620687279

Comment (SergerGood) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:292]: I revert using string interpolation.. but there are many examples. And I'm still in favor of uniform approach. 
https://github.com/dotnet/efcore/blob/ac2bb48b10ecf1289b568a94b7a35e8075c6d787/src/EFCore.Relational/Extensions/RelationalForeignKeyExtensions.cs#L134

https://github.com/dotnet/efcore/blob/ac2bb48b10ecf1289b568a94b7a35e8075c6d787/src/EFCore.Relational/Infrastructure/RelationalOptionsExtension.cs#L451

What about Substring in the same file?

https://github.com/dotnet/efcore/blob/ac2bb48b10ecf1289b568a94b7a35e8075c6d787/src/EFCore.Design/Design/Internal/CSharpHelper.cs#L283

Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs:292]: We can certainly make things more consistent - but given that nothing here is perf-sensitive, I'd go with what produces the shortest/simplest code, which is string interpolation in most cases (e.g. the CommandTimeout case above).

Accordingly, can you revert this particular change as well? All the rest looks fine.

Comment (SergerGood) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs]: Yes. Done

**Review by roji — APPROVED** [review]

File: src/EFCore.Design/Design/Internal/CSharpHelper.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Design/Design/Internal/CSharpHelper.cs]: Thanks.

**Pull request merged**